### DISCUSSION TITLE
forecast price?

### DISCUSSION THREAD

**Discussion by oldwindmill** [amberelectric/public-api#19] [Ideas]
i was using the old api to get the current and the next forecast price, to make decisions on what to do. the new api makes it really easy to get the current price, but to get the forecast i need to return all 96 price intervals and iterate through to find the current and next forecast. 
it would be really nice to have either a 'next forecast' endpoint, or the 'current' price to also return the 'next' price.

or have a missed an easy way to do this?
(also, is the all prices response guaranteed to be in order, first general then feeedin?)-

-russell
ps cool api - i like it....

**Comment by madpilot** [comment]
Let me think about that as a concept. There have already been a bunch of requests around better filtering. Perhaps allowing a iso date and iso time, and doing different things depending on what you ask for.

The order isn't guaranteed. Best to filter for each type separately.

**Reply by oldwindmill** [reply]
Thanks
Is the order guaranteed within a type? (ie, from midnight to midnight?)

**Reply by madpilot** [reply]
Yep, they are ordered by NEM time

**Reply by rob0101** [reply]
Can I suggest adding the next N hours of forecasted prices as an array along with the current price.
Just a simple list of prices.. 

“forecasts” :  [10,11,9,15,33 …….. 21,22]

Users can then slice off however many future forecasted prices as they choose.

I did this with the old API. It mimics the app with the forecasts shown as a list under the current price.

**Reply by madpilot** [reply]
No, I don't want to overload the current price endpoint, given that info can be mapped out of the prices endpoint.

I think tweaking or adding the query parameters to `/prices` is the right direction.

**Reply by oldwindmill** [reply]
i'd be happy with just a parameter to /price to return current+ n forecasts.  or optons on prices for 'now' as the start, and n forecasts. as it is, i've realised i need to get 2 days of data to cover the rollover at midnight.

**Reply by craigcurtin-dev** [reply]
Yeah i am using Node-red and it is pretty easy to slice it using JSONATA based on the Actual/Forecast field -check it out in Node-Red

**Reply by oldwindmill** [reply]
"The order isn't guaranteed. Best to filter for each type separately."
so i cheated...and yesterday the order changed....had to tidy it up anyway.... :)

**Reply by madpilot** [reply]
UPDATE! The order is now guaranteed. It will now always be general > controlled load > feed-in

**Comment by edwapj** [comment]
I am using the old API in a similar way to oldwindmill and rob0101 and would be interested to see what can be done in a similar vein on the new API.
At the half hour plus 90 seconds, my program calls the old API and obtains the actual data for the last 48 periods plus the forecast data for the next 48 or more periods.
The program then extracts the prices for the last completed period, the current period and all of the forecast periods.
It also checks the current price at 5 minute intervals in case price excursions have occurred at short notice.

**Reply by craigcurtin-dev** [reply]
If you already do all that - with the old API - why can't you just do the same with the new API ? (Sorry - just making sure i am not missing something in my code).

I am currently (with the new API in NR) just doing a full call for the whole day (have not tackled the NEM boundary issues yet) and then doing a JSONATA filter on feedin or general to give me two seperate arrays - which i can then cycle through to get the forecast etc going forward

Craig

**Reply by oldwindmill** [reply]
you need to get two days, at least later in the evening as the last get for the day doesnt have a forecast. 
you absolutely can just get the full data set and parse it, but getting 192 data sets when all want is 2 or 3 doesnt seem efficient.
and getting 'today' is easy. tomorrow as well requires working out the date... :(

ps, could you share your jsonata filter? i struggle with it, although i recognise it could be powerful. (i use node red also).

**Reply by madpilot** [reply]
I'm going to add a startTime and endTine query params to the `/prices` endpoint next week

**Reply by edwapj** [reply]
Craig  
A fair question.     It is certainly possible to make a call including the absolute start and finish dates and this is a very flexible and useful option.

However for many applications it would also be straightforward and user-friendly to be able make a relative call and return a sequence of prices indexed to the time of the call.    ( thereby avoiding the need to pass in day/time parameters ).    
This would be similar to the sliding window of data shown on the AEMO dashboard which always shows the most recent 48 actuals plus forecasts whenever it is called.
This simple call would suit many control applications which review the forecast sequence every half hour in order to decide what control action to take for the next half hour.

**Reply by madpilot** [reply]
This is an interesting idea.

The reason the API is designed the way it is is to avoid the overhang problem (because NEM time ends at midnight the next day)

But as for forecast that doesn't matter - you don't care what day it is. Let me play with this as a concept.

**Reply by craigcurtin-dev** [reply]
> you need to get two days, at least later in the evening as the last get for the day doesnt have a forecast.
> you absolutely can just get the full data set and parse it, but getting 192 data sets when all want is 2 or 3 doesnt seem efficient.
> and getting 'today' is easy. tomorrow as well requires working out the date... :(
> 
> ps, could you share your jsonata filter? i struggle with it, although i recognise it could be powerful. (i use node red also).

Here are two Jsonata change nodes - i put them against the feed in data for a day and then break it out into 2 global variables

`[{"id":"f766f49b.2c2868","type":"change","z":"88029ce.abbab6","name":"","rules":[{"t":"set","p":"AmberCost","pt":"global","to":"payload[channelType = \"general\"][[0..50]]","tot":"jsonata"}],"action":"","property":"","from":"","to":"","reg":false,"x":740,"y":200,"wires":[[]]},{"id":"a1506259.e9ced","type":"change","z":"88029ce.abbab6","name":"","rules":[{"t":"set","p":"AmberFeedin","pt":"global","to":"payload[channelType = \"feedIn\"][[0..50]]","tot":"jsonata"}],"action":"","property":"","from":"","to":"","reg":false,"x":750,"y":240,"wires":[[]]}]`

**Reply by madpilot** [reply]
I've added a `previous` and `next` param to the `prices/current` endpoint. You can specify how many intervals you would like.

So to get the current price plus three forecasts, you can do:

```
curl -X 'GET' \
  'https://api.amber.com.au/v1/sites/00000000000000000000000/prices/current?next=3&previous=0' -H 'Authorization: Bearer psk_xxx'
```

Docs have been updated.

**Reply by edwapj** [reply]
Perfecto !!!
Thanks very much
